rewrite-java/src/main/java/org/openrewrite/java/cleanup/UseGenericTypeMethodParameters.java

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright 2022 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java.cleanup;
+
+import org.openrewrite.Cursor;
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Incubating;
+import org.openrewrite.Recipe;
+import org.openrewrite.internal.lang.Nullable;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.JavaTemplate;
+import org.openrewrite.java.JavaVisitor;
+import org.openrewrite.java.MethodMatcher;
+import org.openrewrite.java.search.FindMethods;
+import org.openrewrite.java.tree.*;
+
+import java.time.Duration;
+import java.util.*;
+
+/**
+ * Partial implementation of static analysis RSPEC-3740.
+ * Recipe is scoped to private method declarations, and check the parameterized types passed into the method invocation.
+ * This may be expanded to more use cases when data flow analysis is implemented.
+ */
+@Incubating(since = "7.22.0")
+public class UseGenericTypeMethodParameters extends Recipe {
+
+    @Override
+    public String getDisplayName() {
+        return "Use type parameters instead of raw types";
+    }
+
+    @Override
+    public String getDescription() {
+        return "Replaces raw method parameters on private methods with an inferable generic type. " +
+                "The raw type will not be updated if a raw type or more than 1 type is passed to the method through method invocations in the class. " +
+                "Changes will not be made if a method parameter contains an annotations.";
+    }
+
+    @Override
+    public Set<String> getTags() {
+        return Collections.singleton("RSPEC-3740");
+    }
+
+    @Override
+    public Duration getEstimatedEffortPerOccurrence() {
+        return Duration.ofMinutes(2);
+    }
+
+    public static Set<String> rspecRawTypes = new HashSet<>();
+
+    static {
+        rspecRawTypes.add("java.util.Deque");
+        rspecRawTypes.add("java.util.List");
+        rspecRawTypes.add("java.util.Queue");
+        rspecRawTypes.add("java.util.Set");
+    }
+
+    @Override
+    protected JavaVisitor<ExecutionContext> getVisitor() {
+        return new JavaIsoVisitor<ExecutionContext>() {
+
+            @Override
+            public JavaSourceFile visitJavaSourceFile(JavaSourceFile cu, ExecutionContext executionContext) {
```

Comment:
why this instead of a single-source applicability test?

Reply:
It's mostly to prevent divergence between using UsesType<>("fqn") and the rest of the recipe.

A method declaration parameter and/or each method invocation of the declaration may contain a raw type.
The parameters are checked with `isRawType` which uses the set to qualify what raw types are supported by the recipe.